File: src/Reflector.java
import java.util.HashMap;
import java.util.Map;

public class Reflector {
    private Map<Character, Character> wiring; // Reflector wiring map
    private char name;                        // Reflector identifier

    /*
        Initialize reflector with a specific wiring configuration in the constructor
        @param wiringStr A string representing the reflector wiring
        @param name      Name of the reflector
     */
    public Reflector(String wiringStr, char name) {
        this.wiring = new HashMap<>();
        this.name = name;
        initializeWiring(wiringStr);
    }

    // Initialize reflector wiring based on provided wiring string
    private void initializeWiring(String wiringStr) {
        for (int i = 0; i < wiringStr.length(); i++) {
            char letter = (char)('A' + i);
            char mappedChar = wiringStr.charAt(i);

            // Ensure symmetric mapping
            wiring.put(letter, mappedChar);
            wiring.put(mappedChar, letter);
        }
    }

    // Reflect a character through the reflector wiring (Map character to its pair)
    public char reflect(char inputChar) { return wiring.getOrDefault(inputChar, inputChar); }

    public String display() {
        StringBuilder sb = new StringBuilder();
        sb.append("Reflector").append(name).append(":\n");

        // Display each unique mapping only once
        for (Map.Entry<Character, Character> entry : wiring.entrySet()) {
            char key = entry.getKey();
            char value = entry.getValue();
            if (key < value) // Avoid duplicate pairs
                sb.append("  ").append(key).append(" <-> ").append(value).append("\n");
        }
        return sb.toString();
    }
}

File: web/lib/enigma/reflector.ts
import type { ReflectorName } from "./types";
import { REFLECTOR_WIRINGS } from "./wirings";

export class Reflector {
  readonly name: ReflectorName;
  private readonly wiring: Map<string, string>;

  constructor(name: ReflectorName, wiringStr: string) {
    this.name = name;
    this.wiring = new Map();
    for (let i = 0; i < wiringStr.length; i++) {
      const letter = String.fromCharCode(65 + i);
      const mapped = wiringStr.charAt(i);
      this.wiring.set(letter, mapped);
      this.wiring.set(mapped, letter);
    }
  }

  static fromName(name: ReflectorName): Reflector {
    return new Reflector(name, REFLECTOR_WIRINGS[name]);
  }

  reflect(input: string): string {
    return this.wiring.get(input) ?? input;
  }
}
